WordPress/src/test/java/org/wordpress/android/ui/reader/viewmodels/ReaderViewModelTest.kt

```diff
@@ -0,0 +1,123 @@
+    @Mock lateinit var dateProvider: DateProvider
+    @Mock lateinit var loadReaderTabsUseCase: LoadReaderTabsUseCase
+
+    @Before
+    fun setup() {
+        viewModel = ReaderViewModel(TEST_DISPATCHER, appPrefsWrapper, dateProvider, loadReaderTabsUseCase)
+
+        whenever(dateProvider.getCurrentDate()).thenReturn(Date(DUMMY_CURRENT_TIME))
+    }
+
+    @Test
+    fun `updateTags invoked on first start`() = testWithEmptyTags {
+        // Arrange
+        whenever(appPrefsWrapper.readerTagsUpdatedTimestamp).thenReturn(-1)
+        // Act
+        viewModel.start()
+        // Assert
+        assertThat(viewModel.updateTags.value?.getContentIfNotHandled()).isNotNull
+    }
+
+    @Test
+    fun `updateTags NOT invoked if lastUpdate within threshold`() = testWithEmptyTags {
+        // Arrange
+        whenever(appPrefsWrapper.readerTagsUpdatedTimestamp).thenReturn(DUMMY_CURRENT_TIME - UPDATE_TAGS_THRESHOLD + 1)
+        // Act
+        viewModel.start()
+        // Assert
+        assertThat(viewModel.updateTags.value?.getContentIfNotHandled()).isNull()
+    }
+
+    @Test
+    fun `updateTags invoked if lastUpdate NOT within threshold`() = testWithEmptyTags {
+        // Arrange
+        whenever(appPrefsWrapper.readerTagsUpdatedTimestamp).thenReturn(DUMMY_CURRENT_TIME - UPDATE_TAGS_THRESHOLD - 1)
+        // Act
+        viewModel.start()
+        // Assert
+        assertThat(viewModel.updateTags.value?.getContentIfNotHandled()).isNotNull
+    }
+
+    @Test
+    fun `UiState is NOT updated when loaded tags are empty`() = test {
+        // Arrange
+        whenever(loadReaderTabsUseCase.loadTabs()).thenReturn(ReaderTagList())
+        // Act
+        viewModel.start()
+        // Assert
+        assertThat(viewModel.uiState.value).isNull()
+    }
+
+    @Test
+    fun `UiState is updated in start() when loaded tags are NOT empty`() = test {
+        // Arrange
+        whenever(loadReaderTabsUseCase.loadTabs()).thenReturn(createNonEmptyReaderTagList())
+        var state: ReaderUiState? = null
+        viewModel.uiState.observeForever {
+            state = it
+        }
+        // Act
+        viewModel.start()
+        // Assert
+        assertThat(state).isNotNull
+    }
+
+    @Test
+    fun `Tags are reloaded when FollowedTagsChanged event is received`() = test {
+        // Arrange
+        whenever(loadReaderTabsUseCase.loadTabs()).thenReturn(createNonEmptyReaderTagList())
+        var state: ReaderUiState? = null
+        viewModel.uiState.observeForever {
+            state = it
+        }
+        // Act
+        viewModel.onTagsUpdated(mock())
+        // Assert
+        assertThat(state).isNotNull
+    }
+
+    private fun <T> testWithEmptyTags(block: suspend CoroutineScope.() -> T) {
+        test {
+            whenever(loadReaderTabsUseCase.loadTabs()).thenReturn(ReaderTagList())
+            block()
+        }
+    }
+
+    private fun createNonEmptyReaderTagList(): ReaderTagList {
```

Comment:
What do you think about making this a `testWithNonEmptyTags` function like the one above for consistency?

Reply:
Good point;), fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/11881/commits/88b554dc090c97b5f2a9bd7e4af0d0c702347225